File: android/src/main/java/expo/modules/arcgis/GeometryEngineFunctions.kt
package expo.modules.arcgis

import com.arcgismaps.geometry.Envelope
import com.arcgismaps.geometry.GeodesicEllipseParameters
import com.arcgismaps.geometry.GeodesicSectorParameters
import com.arcgismaps.geometry.Geometry
import com.arcgismaps.geometry.GeometryEngine
import com.arcgismaps.geometry.Multipart
import com.arcgismaps.geometry.Point
import com.arcgismaps.geometry.Polygon
import com.arcgismaps.geometry.Polyline

/**
 * Free functions backing the JS `geometryEngine` namespace. Each decodes its geometry
 * arguments via [geometryFromDict], calls [GeometryEngine] (a Kotlin object), and encodes
 * the result via [dictFromGeometry]. Registered as module `Function`s in `ExpoArcgisModule`.
 */

private fun parseGeo(dict: Map<String, Any?>?): Geometry? = dict?.let { geometryFromDict(it) }
private fun parsePoint(dict: Map<String, Any?>?): Point? = parseGeo(dict) as? Point
private fun encode(g: Geometry?): Map<String, Any?>? = g?.let { dictFromGeometry(it) }
private fun encodeAll(gs: List<Geometry>): List<Map<String, Any?>> = gs.map { dictFromGeometry(it) }

// region Projection & buffer

internal fun geProject(g: Map<String, Any?>, wkid: Int): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.projectOrNull(it, spatialReference(wkid))) }

internal fun geBuffer(g: Map<String, Any?>, distance: Double): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.bufferOrNull(it, distance)) }

internal fun geGeodeticBuffer(
  g: Map<String, Any?>,
  distance: Double,
  unit: String?,
  maxDeviation: Double?,
  curve: String?,
): Map<String, Any?>? = parseGeo(g)?.let {
  encode(GeometryEngine.bufferGeodeticOrNull(it, distance, linearUnit(unit), maxDeviation ?: Double.NaN, curveType(curve)))
}

// region Measurement

internal fun geArea(g: Map<String, Any?>): Double = parseGeo(g)?.let { GeometryEngine.area(it) } ?: 0.0

internal fun geGeodeticArea(g: Map<String, Any?>, unit: String?, curve: String?): Double =
  parseGeo(g)?.let { GeometryEngine.areaGeodetic(it, areaUnit(unit), curveType(curve)) } ?: 0.0

internal fun geLength(g: Map<String, Any?>): Double = parseGeo(g)?.let { GeometryEngine.length(it) } ?: 0.0

internal fun geGeodeticLength(g: Map<String, Any?>, unit: String?, curve: String?): Double =
  parseGeo(g)?.let { GeometryEngine.lengthGeodetic(it, linearUnit(unit), curveType(curve)) } ?: 0.0

internal fun geDistance(a: Map<String, Any?>, b: Map<String, Any?>): Double? {
  val g1 = parseGeo(a) ?: return null
  val g2 = parseGeo(b) ?: return null
  return GeometryEngine.distanceOrNull(g1, g2)
}

internal fun geGeodeticDistance(
  a: Map<String, Any?>,
  b: Map<String, Any?>,
  unit: String?,
  curve: String?,
): Map<String, Any?>? {
  val p1 = parsePoint(a) ?: return null
  val p2 = parsePoint(b) ?: return null
  val result = GeometryEngine.distanceGeodeticOrNull(p1, p2, linearUnit(unit), angularUnit(null), curveType(curve))
    ?: return null
  return mapOf("distance" to result.distance, "azimuth1" to result.azimuth1, "azimuth2" to result.azimuth2)
}

// region Overlay / topological

internal fun geUnion(geometries: List<Map<String, Any?>>): Map<String, Any?>? =
  encode(GeometryEngine.unionOrNull(geometries.mapNotNull { geometryFromDict(it) }))

internal fun geIntersection(a: Map<String, Any?>, b: Map<String, Any?>): Map<String, Any?>? {
  val g1 = parseGeo(a) ?: return null
  val g2 = parseGeo(b) ?: return null
  return encode(GeometryEngine.intersectionOrNull(g1, g2))
}

internal fun geDifference(a: Map<String, Any?>, b: Map<String, Any?>): Map<String, Any?>? {
  val g1 = parseGeo(a) ?: return null
  val g2 = parseGeo(b) ?: return null
  return encode(GeometryEngine.differenceOrNull(g1, g2))
}

internal fun geSymmetricDifference(a: Map<String, Any?>, b: Map<String, Any?>): Map<String, Any?>? {
  val g1 = parseGeo(a) ?: return null
  val g2 = parseGeo(b) ?: return null
  return encode(GeometryEngine.symmetricDifferenceOrNull(g1, g2))
}

internal fun geClip(g: Map<String, Any?>, envelope: Map<String, Any?>): Map<String, Any?>? {
  val geometry = parseGeo(g) ?: return null
  val env = parseGeo(envelope) as? Envelope ?: return null
  return encode(GeometryEngine.clipOrNull(geometry, env))
}

internal fun geCut(g: Map<String, Any?>, cutter: Map<String, Any?>): List<Map<String, Any?>> {
  val geometry = parseGeo(g) ?: return emptyList()
  val line = parseGeo(cutter) as? Polyline ?: return emptyList()
  return encodeAll(GeometryEngine.tryCut(geometry, line))
}

internal fun geConvexHull(g: Map<String, Any?>): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.convexHullOrNull(it)) }

internal fun geLabelPoint(g: Map<String, Any?>): Map<String, Any?>? =
  (parseGeo(g) as? Polygon)?.let { encode(GeometryEngine.labelPointOrNull(it)) }

internal fun geNormalizeCentralMeridian(g: Map<String, Any?>): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.normalizeCentralMeridian(it)) }

internal fun geReshape(g: Map<String, Any?>, reshaper: Map<String, Any?>): Map<String, Any?>? {
  val multipart = parseGeo(g) as? Multipart ?: return null
  val line = parseGeo(reshaper) as? Polyline ?: return null
  return encode(GeometryEngine.reshape(multipart, line))
}

internal fun geIntersections(a: Map<String, Any?>, b: Map<String, Any?>): List<Map<String, Any?>> {
  val g1 = parseGeo(a) ?: return emptyList()
  val g2 = parseGeo(b) ?: return emptyList()
  return GeometryEngine.tryIntersections(g1, g2).mapNotNull { encode(it) }
}

internal fun geExtend(p: Map<String, Any?>, extender: Map<String, Any?>): Map<String, Any?>? {
  val polyline = parseGeo(p) as? Polyline ?: return null
  val ext = parseGeo(extender) as? Polyline ?: return null
  return encode(GeometryEngine.extend(polyline, ext, emptySet()))
}

internal fun geAutoComplete(
  existing: List<Map<String, Any?>>,
  boundaries: List<Map<String, Any?>>,
): List<Map<String, Any?>> {
  val polygons = existing.mapNotNull { parseGeo(it) as? Polygon }
  val lines = boundaries.mapNotNull { parseGeo(it) as? Polyline }
  return GeometryEngine.tryAutoComplete(polygons, lines).mapNotNull { encode(it) }
}

internal fun geBoundary(g: Map<String, Any?>): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.boundaryOrNull(it)) }

internal fun geSimplify(g: Map<String, Any?>): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.simplifyOrNull(it)) }

internal fun geDensify(g: Map<String, Any?>, maxSegmentLength: Double): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.densifyOrNull(it, maxSegmentLength)) }

internal fun geGeneralize(g: Map<String, Any?>, maxDeviation: Double, removeDegenerate: Boolean): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.generalizeOrNull(it, maxDeviation, removeDegenerate)) }

internal fun geOffset(
  g: Map<String, Any?>,
  distance: Double,
  type: String?,
  bevelRatio: Double,
  flattenError: Double,
): Map<String, Any?>? = parseGeo(g)?.let {
  encode(GeometryEngine.offsetOrNull(it, distance, offsetType(type), bevelRatio, flattenError))
}

internal fun geCombineExtents(a: Map<String, Any?>, b: Map<String, Any?>): Map<String, Any?>? {
  val g1 = parseGeo(a) ?: return null
  val g2 = parseGeo(b) ?: return null
  return encode(GeometryEngine.combineExtentsOrNull(g1, g2))
}

// region Relational predicates

internal fun geContains(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.contains(g1, g2)
}
internal fun geCrosses(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.crosses(g1, g2)
}
internal fun geDisjoint(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.disjoint(g1, g2)
}
internal fun geEquals(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.equals(g1, g2)
}
internal fun geIntersects(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.intersects(g1, g2)
}
internal fun geOverlaps(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.overlaps(g1, g2)
}
internal fun geTouches(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.touches(g1, g2)
}
internal fun geWithin(a: Map<String, Any?>, b: Map<String, Any?>): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.within(g1, g2)
}
internal fun geRelate(a: Map<String, Any?>, b: Map<String, Any?>, relation: String): Boolean {
  val g1 = parseGeo(a) ?: return false
  val g2 = parseGeo(b) ?: return false
  return GeometryEngine.relate(g1, g2, relation)
}
internal fun geIsSimple(g: Map<String, Any?>): Boolean = parseGeo(g)?.let { GeometryEngine.isSimple(it) } ?: false

// region Proximity

internal fun geNearestCoordinate(g: Map<String, Any?>, p: Map<String, Any?>): Map<String, Any?>? {
  val geometry = parseGeo(g) ?: return null
  val pt = parsePoint(p) ?: return null
  val result = GeometryEngine.nearestCoordinate(geometry, pt) ?: return null
  return mapOf("coordinate" to dictFromGeometry(result.coordinate), "distance" to result.distance)
}

internal fun geNearestVertex(g: Map<String, Any?>, p: Map<String, Any?>): Map<String, Any?>? {
  val geometry = parseGeo(g) ?: return null
  val pt = parsePoint(p) ?: return null
  val result = GeometryEngine.nearestVertex(geometry, pt) ?: return null
  return mapOf("coordinate" to dictFromGeometry(result.coordinate), "distance" to result.distance)
}

// region Transformations

internal fun geMove(g: Map<String, Any?>, deltaX: Double, deltaY: Double): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.move(it, deltaX, deltaY)) }

internal fun geRotate(g: Map<String, Any?>, angle: Double, origin: Map<String, Any?>?): Map<String, Any?>? {
  val geometry = parseGeo(g) ?: return null
  val pivot = parsePoint(origin)
  return encode(if (pivot != null) GeometryEngine.rotate(geometry, angle, pivot) else GeometryEngine.rotate(geometry, angle))
}

internal fun geScale(
  g: Map<String, Any?>,
  factorX: Double,
  factorY: Double,
  origin: Map<String, Any?>?,
): Map<String, Any?>? {
  val geometry = parseGeo(g) ?: return null
  val pivot = parsePoint(origin)
  return encode(
    if (pivot != null) GeometryEngine.scale(geometry, factorX, factorY, pivot)
    else GeometryEngine.scale(geometry, factorX, factorY),
  )
}

// region Z / M builders

internal fun geWithZ(g: Map<String, Any?>, z: Double): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.createWithZOrNull(it, z)) }

internal fun geWithM(g: Map<String, Any?>, m: Double): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.createWithMOrNull(it, m)) }

internal fun geWithZAndM(g: Map<String, Any?>, z: Double, m: Double): Map<String, Any?>? =
  parseGeo(g)?.let { encode(GeometryEngine.createWithZAndMOrNull(it, z, m)) }

// region Geodesic construction

internal fun geEllipseGeodesic(params: Map<String, Any?>): Map<String, Any?>? {
  val centerDict = params["center"] as? Map<String, Any?> ?: return null
  val center = parsePoint(centerDict) ?: return null

  val semiAxis1Length = (params["semiAxis1Length"] as? Number)?.toDouble() ?: 0.0
  val semiAxis2Length = (params["semiAxis2Length"] as? Number)?.toDouble() ?: 0.0
  val axisDirection   = (params["axisDirection"] as? Number)?.toDouble() ?: 0.0
  val angUnit         = angularUnit(params["angularUnit"] as? String)
  val linUnit         = linearUnit(params["linearUnit"] as? String)
  val maxSegLen       = (params["maxSegmentLength"] as? Number)?.toDouble() ?: 0.0
  val maxPtCount      = (params["maxPointCount"] as? Number)?.toLong() ?: 10L
  val geoType         = params["geometryType"] as? String

  val p: GeodesicEllipseParameters = when (geoType) {
    "polyline"   -> GeodesicEllipseParameters.Companion.createForPolyline()
    "multipoint" -> GeodesicEllipseParameters.Companion.createForMultipoint()
    else         -> GeodesicEllipseParameters.Companion.createForPolygon()
  }
  p.center = center
  p.semiAxis1Length = semiAxis1Length
  p.semiAxis2Length = semiAxis2Length
  p.axisDirection = axisDirection
  p.angularUnit = angUnit
  p.linearUnit = linUnit
  p.maxSegmentLength = maxSegLen
  p.maxPointCount = maxPtCount

  return encode(GeometryEngine.ellipseGeodesicOrNull(p))
}

internal fun geSectorGeodesic(params: Map<String, Any?>): Map<String, Any?>? {
  val centerDict = params["center"] as? Map<String, Any?> ?: return null
  val center = parsePoint(centerDict) ?: return null

  val semiAxis1Length = (params["semiAxis1Length"] as? Number)?.toDouble() ?: 0.0
  val semiAxis2Length = (params["semiAxis2Length"] as? Number)?.toDouble() ?: 0.0
  val axisDirection   = (params["axisDirection"] as? Number)?.toDouble() ?: 0.0
  val sectorAngle     = (params["sectorAngle"] as? Number)?.toDouble() ?: 0.0
  val startDirection  = (params["startDirection"] as? Number)?.toDouble() ?: 0.0
  val angUnit         = angularUnit(params["angularUnit"] as? String)
  val linUnit         = linearUnit(params["linearUnit"] as? String)
  val maxSegLen       = (params["maxSegmentLength"] as? Number)?.toDouble() ?: 0.0
  val maxPtCount      = (params["maxPointCount"] as? Number)?.toLong() ?: 10L
  val geoType         = params["geometryType"] as? String

  val p: GeodesicSectorParameters = when (geoType) {
    "polyline"   -> GeodesicSectorParameters.Companion.createForPolyline()
    "multipoint" -> GeodesicSectorParameters.Companion.createForMultipoint()
    else         -> GeodesicSectorParameters.Companion.createForPolygon()
  }
  p.center = center
  p.semiAxis1Length = semiAxis1Length
  p.semiAxis2Length = semiAxis2Length
  p.axisDirection = axisDirection
  p.sectorAngle = sectorAngle
  p.startDirection = startDirection
  p.angularUnit = angUnit
  p.linearUnit = linUnit
  p.maxSegmentLength = maxSegLen
  p.maxPointCount = maxPtCount

  return encode(GeometryEngine.sectorGeodesicOrNull(p))
}

File: ios/GeometryEngineFunctions.swift
import ArcGIS
import Foundation

/// Free functions backing the JS `geometryEngine` namespace. Each one decodes its
/// geometry arguments via `GeometryCodec`, calls `ArcGIS.GeometryEngine`, and encodes
/// the result back to a JS-friendly value (geometry dict / Double / Bool / array).
/// They are registered as module `Function`s in `ExpoArcgisModule`.

private func parseGeo(_ dict: [String: Any]?) -> Geometry? { dict.flatMap(geometryFromDict) }
private func parsePoint(_ dict: [String: Any]?) -> Point? { parseGeo(dict) as? Point }
private func encode(_ geometry: Geometry?) -> [String: Any]? { geometry.map(dictFromGeometry) }
private func encodeAll(_ geometries: [Geometry]) -> [[String: Any]] { geometries.map(dictFromGeometry) }

// MARK: - Projection & buffer

func geProject(_ g: [String: Any], _ wkid: Int) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.project(geometry, into: spatialReference(wkid: wkid)))
}

func geBuffer(_ g: [String: Any], _ distance: Double) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.buffer(around: geometry, distance: distance))
}

func geGeodeticBuffer(_ g: [String: Any], _ distance: Double, _ unit: String?, _ maxDeviation: Double?, _ curve: String?) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.geodeticBuffer(
    around: geometry, distance: distance, distanceUnit: linearUnit(unit),
    maxDeviation: maxDeviation ?? .nan, curveType: curveType(curve)
  ))
}

// MARK: - Measurement

func geArea(_ g: [String: Any]) -> Double { parseGeo(g).map(GeometryEngine.area(of:)) ?? 0 }

func geGeodeticArea(_ g: [String: Any], _ unit: String?, _ curve: String?) -> Double {
  guard let geometry = parseGeo(g) else { return 0 }
  return GeometryEngine.geodeticArea(of: geometry, unit: areaUnit(unit), curveType: curveType(curve))
}

func geLength(_ g: [String: Any]) -> Double { parseGeo(g).map(GeometryEngine.length(of:)) ?? 0 }

func geGeodeticLength(_ g: [String: Any], _ unit: String?, _ curve: String?) -> Double {
  guard let geometry = parseGeo(g) else { return 0 }
  return GeometryEngine.geodeticLength(of: geometry, lengthUnit: linearUnit(unit), curveType: curveType(curve))
}

func geDistance(_ a: [String: Any], _ b: [String: Any]) -> Double? {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return nil }
  return GeometryEngine.distance(from: g1, to: g2)
}

func geGeodeticDistance(_ a: [String: Any], _ b: [String: Any], _ unit: String?, _ curve: String?) -> [String: Any]? {
  guard let p1 = parsePoint(a), let p2 = parsePoint(b),
    let result = GeometryEngine.geodeticDistance(
      from: p1, to: p2, distanceUnit: linearUnit(unit), azimuthUnit: .degrees, curveType: curveType(curve)
    )
  else { return nil }
  return ["distance": result.distance.value, "azimuth1": result.azimuth1.value, "azimuth2": result.azimuth2.value]
}

// MARK: - Overlay / topological

func geUnion(_ geometries: [[String: Any]]) -> [String: Any]? {
  encode(GeometryEngine.union(of: geometries.compactMap(geometryFromDict)))
}

func geIntersection(_ a: [String: Any], _ b: [String: Any]) -> [String: Any]? {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return nil }
  return encode(GeometryEngine.intersection(g1, g2))
}

func geDifference(_ a: [String: Any], _ b: [String: Any]) -> [String: Any]? {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return nil }
  return encode(GeometryEngine.difference(g1, g2))
}

func geSymmetricDifference(_ a: [String: Any], _ b: [String: Any]) -> [String: Any]? {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return nil }
  return encode(GeometryEngine.symmetricDifference(g1, g2))
}

func geClip(_ g: [String: Any], _ envelope: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g), let env = parseGeo(envelope) as? Envelope else { return nil }
  return encode(GeometryEngine.clip(geometry, to: env))
}

func geCut(_ g: [String: Any], _ cutter: [String: Any]) -> [[String: Any]] {
  guard let geometry = parseGeo(g), let line = parseGeo(cutter) as? Polyline else { return [] }
  return encodeAll(GeometryEngine.cut(geometry, usingCutter: line))
}

func geConvexHull(_ g: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.convexHull(for: geometry))
}

func geLabelPoint(_ g: [String: Any]) -> [String: Any]? {
  guard let polygon = parseGeo(g) as? ArcGIS.Polygon else { return nil }
  return encode(GeometryEngine.labelPoint(for: polygon))
}

func geNormalizeCentralMeridian(_ g: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.normalizeCentralMeridian(of: geometry))
}

func geReshape(_ g: [String: Any], _ reshaper: [String: Any]) -> [String: Any]? {
  guard let line = parseGeo(reshaper) as? ArcGIS.Polyline else { return nil }
  let geometry = parseGeo(g)
  if let polyline = geometry as? ArcGIS.Polyline {
    return encode(GeometryEngine.reshape(polyline, usingReshaper: line))
  }
  if let polygon = geometry as? ArcGIS.Polygon {
    return encode(GeometryEngine.reshape(polygon, usingReshaper: line))
  }
  return nil
}

func geBoundary(_ g: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.boundary(of: geometry))
}

func geSimplify(_ g: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.simplify(geometry))
}

func geDensify(_ g: [String: Any], _ maxSegmentLength: Double) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.densify(geometry, maxSegmentLength: maxSegmentLength))
}

func geGeneralize(_ g: [String: Any], _ maxDeviation: Double, _ removeDegenerate: Bool) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.generalize(geometry, maxDeviation: maxDeviation, removeDegenerateParts: removeDegenerate))
}

func geOffset(_ g: [String: Any], _ distance: Double, _ type: String?, _ bevelRatio: Double, _ flattenError: Double) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.offset(
    geometry, byDistance: distance, offsetType: offsetType(type), bevelRatio: bevelRatio, flattenError: flattenError
  ))
}

func geCombineExtents(_ a: [String: Any], _ b: [String: Any]) -> [String: Any]? {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return nil }
  return encode(GeometryEngine.combineExtents(g1, g2))
}

// MARK: - Relational predicates

func geContains(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.doesGeometry(g1, contain: g2)
}
func geCrosses(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, crossing: g2)
}
func geDisjoint(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, disjointWith: g2)
}
func geEquals(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, equivalentTo: g2)
}
func geIntersects(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, intersecting: g2)
}
func geOverlaps(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, overlapping: g2)
}
func geTouches(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, touching: g2)
}
func geWithin(_ a: [String: Any], _ b: [String: Any]) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, within: g2)
}
func geRelate(_ a: [String: Any], _ b: [String: Any], _ relation: String) -> Bool {
  guard let g1 = parseGeo(a), let g2 = parseGeo(b) else { return false }
  return GeometryEngine.isGeometry(g1, relatedTo: g2, byRelation: relation)
}
func geIsSimple(_ g: [String: Any]) -> Bool { parseGeo(g).map(GeometryEngine.isSimple) ?? false }

// MARK: - Proximity

func geNearestCoordinate(_ g: [String: Any], _ p: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g), let pt = parsePoint(p),
    let result = GeometryEngine.nearestCoordinate(in: geometry, to: pt) else { return nil }
  return ["coordinate": dictFromGeometry(result.coordinate), "distance": result.distance]
}

func geNearestVertex(_ g: [String: Any], _ p: [String: Any]) -> [String: Any]? {
  guard let geometry = parseGeo(g), let pt = parsePoint(p),
    let result = GeometryEngine.nearestVertex(in: geometry, to: pt) else { return nil }
  return ["coordinate": dictFromGeometry(result.coordinate), "distance": result.distance]
}

// MARK: - Transformations

func geMove(_ g: [String: Any], _ deltaX: Double, _ deltaY: Double) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.move(geometry, deltaX: deltaX, deltaY: deltaY))
}

func geRotate(_ g: [String: Any], _ angle: Double, _ origin: [String: Any]?) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.rotate(geometry, by: angle, around: parsePoint(origin)))
}

func geScale(_ g: [String: Any], _ factorX: Double, _ factorY: Double, _ origin: [String: Any]?) -> [String: Any]? {
  guard let geometry = parseGeo(g) else { return nil }
  return encode(GeometryEngine.scale(geometry, factorX: factorX, factorY: factorY, relativeTo: parsePoint(origin)))
}
